File: engine-r/src/main/java/pl/psi/aaas/engine/r/transceiver/RTransceiverFactory.java
package pl.psi.aaas.engine.r.transceiver;

import org.rosuda.REngine.Rserve.RConnection;
import pl.psi.aaas.DataFrame;
import pl.psi.aaas.Parameter;
import pl.psi.aaas.Primitive;
import pl.psi.aaas.Vector;
import pl.psi.aaas.engine.r.RValuesTransceiver;
import pl.psi.aaas.usecase.CalculationDefinitionIf;
import pl.psi.aaas.usecase.CalculationException;

import java.time.ZonedDateTime;

public class RTransceiverFactory {
    public static RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf > get(Parameter<?> param, RConnection conn) throws CalculationException {
        if (param instanceof Primitive) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) primitiveTransceiver(((Primitive<?>) param).getClazz(), conn);
        } else if (param instanceof Vector) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) vectorTransceiver((Vector<?>) param, conn);
        } else if (param instanceof DataFrame) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) new DataFrameTransceiver(conn);
        }
        throw new CalculationException("Unsupported parameter type: " + param.getClass());
    }

    private static RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf > primitiveTransceiver(Class<?> clazz, RConnection conn) throws CalculationException {
        if (clazz == java.lang.String.class || clazz == String.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RPrimitiveTransceiver.string(conn);
        } else if (clazz == java.lang.Long.class || clazz == Long.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RPrimitiveTransceiver.longTransceiver(conn);
        } else if (clazz == java.lang.Double.class || clazz == Double.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RPrimitiveTransceiver.doubleTransceiver(conn);
        } else if (clazz == java.lang.Boolean.class || clazz == Boolean.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RPrimitiveTransceiver.booleanTransceiver(conn);
        } else if (clazz == ZonedDateTime.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) new DateTimeTransceiver(conn);
        } else {
            throw new CalculationException("Not implemented parameter type " + clazz);
        }
    }

    private static RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf > vectorTransceiver(Vector<?> param, RConnection conn) throws CalculationException {
        Class<?> elemClazz = param.getElemClazz();
        if (elemClazz == java.lang.String.class || elemClazz == String.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RArrayTransceiver.string(conn);
        } else if (elemClazz == java.lang.Long.class || elemClazz == Long.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RArrayTransceiver.longTransceiver(conn);
        } else if (elemClazz == java.lang.Double.class || elemClazz == Double.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RArrayTransceiver.doubleTransceiver(conn);
        } else if (elemClazz == java.lang.Boolean.class || elemClazz == Boolean.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) RArrayTransceiver.booleanTransceiver(conn);
        } else if (elemClazz == ZonedDateTime.class) {
            return (RValuesTransceiver<Parameter<?>, ?, CalculationDefinitionIf >) (Object) new ArrayDateTimeTransceiver(conn);
        } else {
            throw new CalculationException("Not implemented array parameter type " + elemClazz);
        }
    }
}

File: engine-r/src/main/kotlin/pl/psi/aaas/engine/r/transceiver/RTransceiverFactory.kt
package pl.psi.aaas.engine.r.transceiver

import org.rosuda.REngine.Rserve.RConnection
import pl.psi.aaas.engine.r.RValuesTransceiver
import pl.psi.aaas.usecase.CalculationDefinition
import pl.psi.aaas.usecase.CalculationException
import pl.psi.aaas.DataFrame
import pl.psi.aaas.Parameter
import pl.psi.aaas.Primitive
import pl.psi.aaas.Vector
import java.time.ZonedDateTime

// TODO 09.05.2018 kskitek: this factory has to be generic in terms ofPrimitive Engine impl; provided separately; registration ofPrimitive impls with SPI
// TODO rather than using instanceof and casting use visitor/handler as a chain of first
internal object RValuesTransceiverFactory {
    fun get(param: Parameter<*>, conn: RConnection): RValuesTransceiver<Parameter<*>, *, CalculationDefinition> =
            when (param) {
                is Primitive -> primitiveTransceiver(param.clazz as Class<Any>, conn)
                is Vector<*> -> vectorTransceiver(param, conn)
                is DataFrame -> DataFrameTransceiver(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            }

    private fun primitiveTransceiver(clazz: Class<Any>, conn: RConnection): RValuesTransceiver<Parameter<*>, *, CalculationDefinition> =
            when (clazz) {
                java.lang.String::class.java  -> RPrimitiveTransceiver.string(conn)
                String::class.java            -> RPrimitiveTransceiver.string(conn)
                java.lang.Long::class.java    -> RPrimitiveTransceiver.long(conn)
                Long::class.java              -> RPrimitiveTransceiver.long(conn)
                java.lang.Double::class.java  -> RPrimitiveTransceiver.double(conn)
                Double::class.java            -> RPrimitiveTransceiver.double(conn)
                java.lang.Boolean::class.java -> RPrimitiveTransceiver.boolean(conn)
                Boolean::class.java           -> RPrimitiveTransceiver.boolean(conn)
                ZonedDateTime::class.java     -> DateTimeTransceiver(conn) as RValuesTransceiver<*, *, *>
                else                          -> throw CalculationException("Not implemented parameter type $clazz")
            } as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
}

private fun vectorTransceiver(param: Vector<*>, conn: RConnection): RValuesTransceiver<Parameter<*>, *, CalculationDefinition> = //EngineValuesSender TODO
        when (param.elemClazz) {
            java.lang.String::class.java  -> RArrayTransceiver.string(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            String::class.java            -> RArrayTransceiver.string(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            java.lang.Long::class.java    -> RArrayTransceiver.long(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            Long::class.java              -> RArrayTransceiver.long(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            java.lang.Double::class.java  -> RArrayTransceiver.double(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            Double::class.java            -> RArrayTransceiver.double(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            java.lang.Boolean::class.java -> RArrayTransceiver.boolean(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            Boolean::class.java           -> RArrayTransceiver.boolean(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            ZonedDateTime::class.java     -> ArrayDateTimeTransceiver(conn) as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>
            else                          -> throw CalculationException("Not implemented array parameter type ${param.elemClazz}")
        } as RValuesTransceiver<Parameter<*>, *, CalculationDefinition>

